### DISCUSSION TITLE
Putting code into ram doesn't help?

### DISCUSSION THREAD

**Discussion by Huckies** [hathach/tinyusb#1557] [Q&A]
Hi there, I'm implementing a composite(multi interface) HID device on ESP32-S3.

I found this in one of the examples:
```
void hid_task(void)
{
  // Poll every 10ms
  const uint32_t interval_ms = 10;
  static uint32_t start_ms = 0;

  if ( board_millis() - start_ms < interval_ms) return; // not enough time
  start_ms += interval_ms;

  uint32_t const btn = board_button_read();

  // Remote wakeup
  if ( tud_suspended() && btn )
  {
    // Wake up host if we are in suspend mode
    // and REMOTE_WAKEUP feature is enabled by host
    tud_remote_wakeup();
  }

  /*------------- Keyboard -------------*/
  if ( tud_hid_n_ready(ITF_KEYBOARD) )
  {
    // use to avoid send multiple consecutive zero report for keyboard
    static bool has_key = false;

    if ( btn )
    {
      uint8_t keycode[6] = { 0 };
      keycode[0] = HID_KEY_A;

      tud_hid_n_keyboard_report(ITF_KEYBOARD, 0, 0, keycode);

      has_key = true;
    }else
    {
      // send empty key report if previously has key pressed
      if (has_key) tud_hid_n_keyboard_report(ITF_KEYBOARD, 0, 0, NULL);
      has_key = false;
    }
  }

  /*------------- Mouse -------------*/
  if ( tud_hid_n_ready(ITF_MOUSE) )
  {
    if ( btn )
    {
      int8_t const delta = 5;

      // no button, right + down, no scroll pan
      tud_hid_n_mouse_report(ITF_MOUSE, 0, 0x00, delta, delta, 0, 0);
    }
  }
}
```
Seems it checks the availability of each interface(actually it's IN endpoint) and then send keyboard AND mouse reports in one polling interval.

But when it comes to me the usb interface can stop response if I move mouse rapidly while typing fast on the keyboard. Turning to send either a keyboard report or a mouse report per polling interval fixes this.

All the difference is that I put ``hid_task()`` into iram and use a FreeRTOS API ``vTaskDelayUntil()`` instead of 
```
  const uint32_t interval_ms = 10;
  static uint32_t start_ms = 0;

  if ( board_millis() - start_ms < interval_ms) return; // not enough time
  start_ms += interval_ms;
```

Can ram code be reason that leads to a "delayed" status update and let the stack mistaken a busy endpoint as a free one?

**Comment by pigrew** [comment]
Since there are two separate interfaces (one for keyboard and one for mouse), they shouldn't interact with each other. A report can be pending in both interrupt-IN endpoints simultaneously. (I assume the host is bug-free).

It sounds like you have a much more complicated example of a composite device with a keyboard and mouse, and also that you are using FreeRTOS?

You need to verify that `tud_task` is called sufficiently often, or you use the FreeRTOS mode of TinyUSB where TinyUSB has its own task.

Forcing the code into RAM shouldn't change anything... but could expose race conditions in multithreaded applications.

As it is now, the `usbd_edpt_xfer` function (called by the classes) is not reentrant... it has race conditions if you perform USB operations from multiple tasks simultaneously on a single endpoint.

**Reply by Huckies** [reply]
Thank you for the rapid reply, ``tud_task()`` is called by a FreeRTOS task with higher priority than ``hid_task()``. I also checked to make sure that ``hid_task()`` is the only task I implemented to call ``usbd_edpt_xfer()`` .

Would you please introduce more on `the FreeRTOS mode of TinyUSB where TinyUSB has its own task`?

**Reply by Huckies** [reply]
Further check indicates that keyboard endpoint alone stuck after a few transmissions, the report fetched from it is truncated and the endpoint it self runs into a 'never ready' state.

**Comment by hathach** [comment]
that example is a bit old, I would suggest to use a report queue/buffer + tud_hid_report_complete_cb() to prevent race condition.

https://github.com/hathach/tinyusb/blob/master/examples/device/hid_composite_freertos/src/main.c
